File: claude-log-tail.py
#!/usr/bin/env python3
from __future__ import annotations
"""
claude-log-tail.py - Monitor Claude JSONL conversation logs with colored output

Usage: ./claude-log-tail.py <directory>

No external dependencies required - uses Python stdlib only.
"""

import json
import os
import signal
import sys
import time
from datetime import datetime, timezone

# ANSI colors
C_RESET = '\033[0m'
C_TIME = '\033[38;5;243m'
C_PROJ = '\033[38;5;33m'
C_USER = '\033[38;5;34m'
C_ASST = '\033[38;5;208m'
C_TOOL = '\033[38;5;141m'
C_ERR = '\033[38;5;196m'
C_DEF = '\033[38;5;252m'

MAX_CONTENT_LEN = 300
POLL_INTERVAL = 0.5  # Check for changes every 0.5 seconds
FILE_SCAN_INTERVAL = 10  # Check for new files every 10 seconds


def get_project_name(filepath: str, basedir: str) -> str:
    """Extract clean project name from file path."""
    rel = os.path.relpath(filepath, basedir)
    project = rel.split(os.sep)[0]

    home = os.path.expanduser("~")
    home_prefix = home.lstrip("/").replace("/", "-") + "-"

    if project.startswith("-"):
        project = project[1:]
    if project.startswith(home_prefix):
        project = project[len(home_prefix):]

    return project


def extract_content(msg_content) -> str:
    """Extract readable content from message."""
    if isinstance(msg_content, str):
        return msg_content
    elif isinstance(msg_content, list):
        parts = []
        for item in msg_content:
            t = item.get("type", "")
            if t == "text":
                parts.append(item.get("text", ""))
            elif t == "thinking":
                parts.append("[thinking]")
            elif t == "tool_use":
                name = item.get("name", "?")
                inp = item.get("input", {})
                # Show useful snippet based on tool type
                if name == "Bash" and inp.get("command"):
                    cmd = inp["command"][:80].replace("\n", " ")
                    parts.append(f"$ {cmd}")
                elif name == "Read" and inp.get("file_path"):
                    parts.append(f"read {inp['file_path']}")
                elif name == "Edit" and inp.get("file_path"):
                    parts.append(f"edit {inp['file_path']}")
                elif name == "Write" and inp.get("file_path"):
                    parts.append(f"write {inp['file_path']}")
                elif name == "Glob" and inp.get("pattern"):
                    parts.append(f"glob {inp['pattern']}")
                elif name == "Grep" and inp.get("pattern"):
                    parts.append(f"grep {inp['pattern'][:50]}")
                elif name == "Task" and inp.get("prompt"):
                    parts.append(f"task: {inp['prompt'][:60]}")
                else:
                    parts.append(f"[{name}]")
            elif t == "tool_result":
                content = item.get("content", "")
                s = content if isinstance(content, str) else str(content)
                parts.append(s[:100])
        return " | ".join(p for p in parts if p)
    return ""


def get_effective_type(data: dict) -> str:
    """Determine effective message type."""
    msg_type = data.get("type", "")
    msg_content = data.get("message", {}).get("content")

    if msg_type == "user" and isinstance(msg_content, list):
        if any(item.get("type") == "tool_result" for item in msg_content):
            return "tool"
    return msg_type


def parse_timestamp_local(ts: str) -> str:
    """Parse ISO timestamp and return local time HH:MM:SS."""
    if not ts or "T" not in ts:
        return ""
    try:
        # Handle various ISO formats
        ts_clean = ts.replace("Z", "+00:00")
        if "." in ts_clean:
            # Truncate microseconds if too long
            parts = ts_clean.split(".")
            frac_and_tz = parts[1]
            # Find where timezone starts (+ or -)
            for i, c in enumerate(frac_and_tz):
                if c in "+-":
                    frac = frac_and_tz[:i][:6]  # max 6 digits
                    tz = frac_and_tz[i:]
                    ts_clean = f"{parts[0]}.{frac}{tz}"
                    break
            else:
                ts_clean = f"{parts[0]}.{frac_and_tz[:6]}"
        dt = datetime.fromisoformat(ts_clean)
        # Convert to local time
        local_dt = dt.astimezone()
        return local_dt.strftime("%H:%M:%S")
    except (ValueError, AttributeError):
        # Fallback: just extract time portion
        return ts.split("T")[1].split(".")[0] if "T" in ts else ""


def format_line(line: str, project: str) -> str | None:
    """Parse JSON line and format for display."""
    try:
        data = json.loads(line)
    except json.JSONDecodeError:
        return None

    msg_type = get_effective_type(data)
    if msg_type in ("file-history-snapshot", "progress"):
        return None

    ts = data.get("timestamp", "")
    time_part = parse_timestamp_local(ts)

    msg = data.get("message", {})
    content = extract_content(msg.get("content", ""))
    if not content:
        return None

    content = " ".join(content.replace("\n", " ").split())
    if len(content) > MAX_CONTENT_LEN:
        content = content[:MAX_CONTENT_LEN] + "..."

    colors = {"user": C_USER, "assistant": C_ASST, "tool": C_TOOL}
    color = colors.get(msg_type, C_DEF)
    if "error" in content.lower():
        color = C_ERR

    return f"{C_TIME}[{time_part}]{C_RESET} {C_PROJ}[{project}]{C_RESET} {color}[{msg_type}]{C_RESET} {content}"


def find_jsonl_files(basedir: str) -> set:
    """Find all .jsonl files under basedir."""
    files = set()
    for root, _, filenames in os.walk(basedir):
        for fname in filenames:
            if fname.endswith(".jsonl"):
                files.add(os.path.join(root, fname))
    return files


def signal_handler(signum, frame):
    """Handle shutdown signals."""
    print(f"\n{C_TIME}Shutting down...{C_RESET}", flush=True)
    os._exit(0)


def main():
    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} <directory>")
        sys.exit(1)

    basedir = os.path.abspath(sys.argv[1])
    if not os.path.isdir(basedir):
        print(f"Error: Directory does not exist: {basedir}", file=sys.stderr)
        sys.exit(1)

    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)

    print(f"Monitoring JSONL files in: {basedir}")
    print(f"Polling every {POLL_INTERVAL}s. New files checked every {FILE_SCAN_INTERVAL}s.")
    print("Press Ctrl+C to exit.")
    print("---")

    # Track file positions: filepath -> (size, position)
    file_positions: dict[str, int] = {}

    known_files = find_jsonl_files(basedir)
    print(f"Monitoring {len(known_files)} JSONL files")
    print("---")

    # Initialize positions to end of files (don't show historical content)
    for filepath in known_files:
        try:
            file_positions[filepath] = os.path.getsize(filepath)
        except OSError:
            file_positions[filepath] = 0

    last_file_scan = time.time()

    while True:
        # Check each file for new content
        for filepath in list(known_files):
            try:
                current_size = os.path.getsize(filepath)
                last_pos = file_positions.get(filepath, 0)

                if current_size > last_pos:
                    # File has grown - read new content
                    project = get_project_name(filepath, basedir)
                    with open(filepath, "r", encoding="utf-8", errors="replace") as f:
                        f.seek(last_pos)
                        for line in f:
                            line = line.rstrip()
                            if line:
                                formatted = format_line(line, project)
                                if formatted:
                                    print(formatted, flush=True)
                        file_positions[filepath] = f.tell()

            except OSError:
                # File may have been deleted
                pass

        # Periodically check for new files
        now = time.time()
        if now - last_file_scan >= FILE_SCAN_INTERVAL:
            last_file_scan = now
            current_files = find_jsonl_files(basedir)
            new_files = current_files - known_files

            if new_files:
                for f in sorted(new_files):
                    project = get_project_name(f, basedir)
                    print(f"{C_TIME}[{time.strftime('%H:%M:%S')}]{C_RESET} {C_PROJ}[+]{C_RESET} {project}")
                    # Start at end of new file (don't show historical content)
                    try:
                        file_positions[f] = os.path.getsize(f)
                    except OSError:
                        file_positions[f] = 0

                known_files = current_files

            # Also clean up deleted files
            deleted = known_files - current_files
            for f in deleted:
                file_positions.pop(f, None)
            known_files = current_files

        time.sleep(POLL_INTERVAL)


if __name__ == "__main__":
    main()

File: claude-log-tail.js
#!/usr/bin/env node
"use strict";

/**
 * claude-log-tail.js - Monitor Claude JSONL conversation logs with colored output
 *
 * Usage: claude-log-tail <directory>
 *
 * No external dependencies required - uses Node.js stdlib only.
 */

const fs = require("fs");
const os = require("os");
const path = require("path");

const C_RESET = "\x1b[0m";
const C_TIME = "\x1b[38;5;243m";
const C_PROJ = "\x1b[38;5;33m";
const C_USER = "\x1b[38;5;34m";
const C_ASST = "\x1b[38;5;208m";
const C_TOOL = "\x1b[38;5;141m";
const C_ERR = "\x1b[38;5;196m";
const C_DEF = "\x1b[38;5;252m";

const MAX_CONTENT_LEN = 300;
const POLL_INTERVAL_MS = 5000;
const DRAIN_DELAY_MS = 50;

function getProjectName(filepath, basedir) {
  const rel = path.relative(basedir, filepath);
  let project = rel.split(path.sep)[0] || rel;

  const home = os.homedir();
  const homePrefix = home.replace(/^\/+/, "").split(path.sep).join("-") + "-";

  if (project.startsWith("-")) {
    project = project.slice(1);
  }
  if (project.startsWith(homePrefix)) {
    project = project.slice(homePrefix.length);
  }

  return project;
}

function extractContent(messageContent) {
  if (typeof messageContent === "string") {
    return messageContent;
  }

  if (!Array.isArray(messageContent)) {
    return "";
  }

  const parts = [];
  for (const item of messageContent) {
    if (!item || typeof item !== "object") {
      continue;
    }

    const type = item.type || "";
    if (type === "text") {
      parts.push(item.text || "");
      continue;
    }

    if (type === "thinking") {
      parts.push("[thinking]");
      continue;
    }

    if (type === "tool_use") {
      const name = item.name || "?";
      const input = item.input && typeof item.input === "object" ? item.input : {};

      if (name === "Bash" && input.command) {
        const command = String(input.command).slice(0, 80).replace(/\n/g, " ");
        parts.push(`$ ${command}`);
      } else if (name === "Read" && input.file_path) {
        parts.push(`read ${input.file_path}`);
      } else if (name === "Edit" && input.file_path) {
        parts.push(`edit ${input.file_path}`);
      } else if (name === "Write" && input.file_path) {
        parts.push(`write ${input.file_path}`);
      } else if (name === "Glob" && input.pattern) {
        parts.push(`glob ${input.pattern}`);
      } else if (name === "Grep" && input.pattern) {
        parts.push(`grep ${String(input.pattern).slice(0, 50)}`);
      } else if (name === "Task" && input.prompt) {
        parts.push(`task: ${String(input.prompt).slice(0, 60)}`);
      } else {
        parts.push(`[${name}]`);
      }
      continue;
    }

    if (type === "tool_result") {
      const content = item.content;
      const text = typeof content === "string" ? content : String(content);
      parts.push(text.slice(0, 100));
    }
  }

  return parts.filter(Boolean).join(" | ");
}

function getEffectiveType(data) {
  const messageType = data.type || "";
  const messageContent = data.message && data.message.content;

  if (messageType === "user" && Array.isArray(messageContent)) {
    if (messageContent.some((item) => item && item.type === "tool_result")) {
      return "tool";
    }
  }

  return messageType;
}

function parseTimestampLocal(ts) {
  if (!ts || !ts.includes("T")) {
    return "";
  }

  try {
    let tsClean = ts.replace("Z", "+00:00");

    const dotIndex = tsClean.indexOf(".");
    if (dotIndex >= 0) {
      const base = tsClean.slice(0, dotIndex);
      const fracAndTz = tsClean.slice(dotIndex + 1);
      const tzMatch = fracAndTz.match(/[+-]/);

      if (tzMatch && tzMatch.index !== undefined) {
        const idx = tzMatch.index;
        const frac = fracAndTz.slice(0, idx).slice(0, 6);
        const tz = fracAndTz.slice(idx);
        tsClean = `${base}.${frac}${tz}`;
      } else {
        tsClean = `${base}.${fracAndTz.slice(0, 6)}`;
      }
    }

    const dt = new Date(tsClean);
    if (Number.isNaN(dt.getTime())) {
      throw new Error("Invalid date");
    }

    return dt.toLocaleTimeString("en-US", {
      hour12: false,
      hour: "2-digit",
      minute: "2-digit",
      second: "2-digit",
    });
  } catch {
    if (!ts.includes("T")) {
      return "";
    }
    return ts.split("T")[1].split(".")[0];
  }
}

function formatLine(line, project) {
  let data;
  try {
    data = JSON.parse(line);
  } catch {
    return null;
  }

  const messageType = getEffectiveType(data);
  if (messageType === "file-history-snapshot" || messageType === "progress") {
    return null;
  }

  const timePart = parseTimestampLocal(data.timestamp || "");
  const message = data.message && typeof data.message === "object" ? data.message : {};
  let content = extractContent(message.content);
  if (!content) {
    return null;
  }

  content = content.replace(/\n/g, " ").split(/\s+/).filter(Boolean).join(" ");
  if (content.length > MAX_CONTENT_LEN) {
    content = `${content.slice(0, MAX_CONTENT_LEN)}...`;
  }

  const colors = {
    user: C_USER,
    assistant: C_ASST,
    tool: C_TOOL,
  };
  let color = colors[messageType] || C_DEF;
  if (content.toLowerCase().includes("error")) {
    color = C_ERR;
  }

  return `${C_TIME}[${timePart}]${C_RESET} ${C_PROJ}[${project}]${C_RESET} ${color}[${messageType}]${C_RESET} ${content}`;
}

function findJsonlFiles(basedir) {
  const files = new Set();
  const stack = [basedir];

  while (stack.length > 0) {
    const current = stack.pop();
    if (!current) {
      continue;
    }

    let entries;
    try {
      entries = fs.readdirSync(current, { withFileTypes: true });
    } catch {
      continue;
    }

    for (const entry of entries) {
      const fullPath = path.join(current, entry.name);
      if (entry.isDirectory()) {
        stack.push(fullPath);
      } else if (entry.isFile() && entry.name.endsWith(".jsonl")) {
        files.add(fullPath);
      }
    }
  }

  return files;
}

function drainFile(filepath, filePositions, basedir) {
  try {
    const currentSize = fs.statSync(filepath).size;
    const lastPos = filePositions.get(filepath) || 0;

    if (currentSize <= lastPos) {
      return;
    }

    const project = getProjectName(filepath, basedir);
    const bytesToRead = currentSize - lastPos;
    const fd = fs.openSync(filepath, "r");

    try {
      const buffer = Buffer.alloc(bytesToRead);
      const bytesRead = fs.readSync(fd, buffer, 0, bytesToRead, lastPos);
      const text = buffer.toString("utf8", 0, bytesRead);

      for (const lineRaw of text.split(/\r?\n/)) {
        const line = lineRaw.trimEnd();
        if (!line) {
          continue;
        }

        const formatted = formatLine(line, project);
        if (formatted) {
          process.stdout.write(`${formatted}\n`);
        }
      }

      filePositions.set(filepath, lastPos + bytesRead);
    } finally {
      fs.closeSync(fd);
    }
  } catch {
    // File may have been deleted
  }
}

function run() {
  if (process.argv.length < 3) {
    process.stdout.write("Usage: claude-log-tail <directory>\n");
    process.exit(1);
  }

  const basedir = path.resolve(process.argv[2]);
  if (!fs.existsSync(basedir) || !fs.statSync(basedir).isDirectory()) {
    console.error(`Error: Directory does not exist: ${basedir}`);
    process.exit(1);
  }

  const shutdown = () => {
    if (rootWatcher) {
      rootWatcher.close();
    }
    for (const watcher of dirWatchers.values()) {
      try {
        watcher.close();
      } catch {
        // ignore
      }
    }
    clearInterval(pollTimer);
    process.stdout.write(`\n${C_TIME}Shutting down...${C_RESET}\n`);
    process.exit(0);
  };
  process.on("SIGINT", shutdown);
  process.on("SIGTERM", shutdown);

  const filePositions = new Map();
  const dirWatchers = new Map();

  // --- Debounced drain mechanism ---
  const changedFiles = new Set();
  let drainScheduled = false;

  function scheduleDrain() {
    if (drainScheduled) {
      return;
    }
    drainScheduled = true;
    setTimeout(flushChanges, DRAIN_DELAY_MS);
  }

  function flushChanges() {
    drainScheduled = false;
    const paths = Array.from(changedFiles);
    changedFiles.clear();

    for (const filepath of paths) {
      if (!filePositions.has(filepath)) {
        continue;
      }
      drainFile(filepath, filePositions, basedir);
    }
  }

  // --- Per-directory watcher ---
  function watchDirectory(dirPath) {
    if (dirWatchers.has(dirPath)) {
      return;
    }

    try {
      const watcher = fs.watch(dirPath, (_eventType, filename) => {
        if (!filename) {
          return;
        }

        const fullPath = path.join(dirPath, filename);

        // Already tracked? Mark changed.
        if (filePositions.has(fullPath)) {
          changedFiles.add(fullPath);
          scheduleDrain();
          return;
        }

        // New .jsonl file?
        if (filename.endsWith(".jsonl")) {
          let stats;
          try {
            stats = fs.statSync(fullPath);
          } catch {
            return;
          }
          if (!stats.isFile()) {
            return;
          }

          filePositions.set(fullPath, 0);
          watchDirectory(path.dirname(fullPath));

          const project = getProjectName(fullPath, basedir);
          const time = new Date().toLocaleTimeString("en-US", {
            hour12: false,
            hour: "2-digit",
            minute: "2-digit",
            second: "2-digit",
          });
          process.stdout.write(`${C_TIME}[${time}]${C_RESET} ${C_PROJ}[+]${C_RESET} ${project}\n`);

          changedFiles.add(fullPath);
          scheduleDrain();
        }
      });

      watcher.on("error", () => {
        dirWatchers.delete(dirPath);
        try {
          watcher.close();
        } catch {
          // ignore
        }
      });

      dirWatchers.set(dirPath, watcher);
    } catch {
      // Can't watch this directory; will rely on fallback poll
    }
  }

  // --- Recursive watcher on root for new file/directory discovery ---
  // On macOS, fs.watch({ recursive: true }) uses FSEvents — a single
  // efficient kernel subscription. We filter early: only react to .jsonl
  // filenames, skipping costly statSync on everything else.
  let rootWatcher = null;
  try {
    rootWatcher = fs.watch(basedir, { recursive: true }, (_eventType, filename) => {
      if (!filename) {
        return;
      }

      const basename = path.basename(filename);
      const fullPath = path.join(basedir, filename);

      // Already tracked? Mark changed.
      if (filePositions.has(fullPath)) {
        changedFiles.add(fullPath);
        scheduleDrain();
        return;
      }

      // Only stat .jsonl files — skip everything else.
      if (!basename.endsWith(".jsonl")) {
        return;
      }

      let stats;
      try {
        stats = fs.statSync(fullPath);
      } catch {
        return;
      }
      if (!stats.isFile()) {
        return;
      }

      // New .jsonl file discovered
      filePositions.set(fullPath, 0);
      watchDirectory(path.dirname(fullPath));

      const project = getProjectName(fullPath, basedir);
      const time = new Date().toLocaleTimeString("en-US", {
        hour12: false,
        hour: "2-digit",
        minute: "2-digit",
        second: "2-digit",
      });
      process.stdout.write(`${C_TIME}[${time}]${C_RESET} ${C_PROJ}[+]${C_RESET} ${project}\n`);

      changedFiles.add(fullPath);
      scheduleDrain();
    });

    rootWatcher.on("error", () => {
      if (rootWatcher) {
        rootWatcher.close();
        rootWatcher = null;
      }
    });
  } catch {
    rootWatcher = null;
  }

  // --- Initial file discovery ---
  process.stdout.write(`Monitoring JSONL files in: ${basedir}\n`);

  const knownFiles = findJsonlFiles(basedir);
  for (const filepath of knownFiles) {
    try {
      filePositions.set(filepath, fs.statSync(filepath).size);
    } catch {
      filePositions.set(filepath, 0);
    }
    watchDirectory(path.dirname(filepath));
  }

  process.stdout.write(
    `Monitoring ${filePositions.size} JSONL files in ${dirWatchers.size} directories.\n`,
  );
  process.stdout.write("Press Ctrl+C to exit.\n");
  process.stdout.write("---\n");

  // --- Fallback poll: stat only tracked files (no full tree walk) ---
  function pollTrackedFiles() {
    for (const filepath of filePositions.keys()) {
      drainFile(filepath, filePositions, basedir);
    }
  }

  const pollTimer = setInterval(pollTrackedFiles, POLL_INTERVAL_MS);
}

run();
